ipc_service: ipc_icbmsg: Rebounding functionality

rakons opened the pull request:
This PR covers many changes inside ICBMsg backend implementation that allows real rebonding. The side effect of that changes is different mechanism of memory allocation where both sides shares the information about the used memory. This is part of the functionality of channel rebounding after one of the communication side is restarted.

Note that bitpool is implemented and the usage of bitarray is replaced by bitpool. The bitpool is expansion of atomic subsystem, not replacement for bitarray.

Some side work was done to ipc_sessions test for clearer configuration and to fix some issues found during work.

rakons committed:
ipc: bitpool: Implementation

Implementation of the atomic bitpool library that allows to prepare any
operation of bit arrays and storing it atomically using copy-cas approach.

The bitpool besides of some similarities to bitarray is designed to rather
expand atomic module with its atomic bit arrays than bitarray module.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
ipc_service: ipc_icbmsg: New buffer allocation scheme

Information about buffer usage is now placed in shared memory
used for communication.
This is basic requirement to restore communication after one
core is restarted.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
ipc_service: ipc_icbmsg: Real unbound implementation

Remove old unbound implementation that was rather holding the endpoint
and implement new one that really removes the endpoint bound.
The remote is now informed about endpoint unbounding and unbounded
endpoint returns to unconfigured state.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
ipc_service: ipc_icbmsg: Safe control channel bounding

Change the way the control channel is established.
The connection process is moved from instance opening to
first endpoint bounding.
This way, even if the control channel is unbounded, the reconnection
process would be started as soon as user request new endpoint bounding.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
tests: ipc_sessions: Wait for buffer in test_tx_long

There may be a situation where after a bunch of long transfers
transmitter buffer is not available to send RXGET command.
Address such a situation in a test to allow the test to PASS if
the buffer is available after a short delay.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
tests: ipc_sessions: Small fixes

More log messages.
Fix REBOOT log to REBOND in the REBOND command execution.
Do not pause WDT while sleeping during remote core restart.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
tests: ipc_sessions: Add ICBMsg test configuration for nRF5340DK.

ICBMSG test configuration for nRF5340DK.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
tests: ipc_sessions: Add packed attribute to test_cmd_xdata

Make sure that test_cmt_xdata structure is aligned as expected
independently from test block size and architecture.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
tests: ipc_sessions: Use icmsg suffix for nrf54h20dk_cpuapp_cpurad test

Use icmsg file suffix for nrf54h20dk_cpuapp_cpurad test to align it with
other tests.
ICMSG used as IPC backend in communication between cpuapp and cpurad
is not a default configuration for the board.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
tests: ipc_sessions: Add ICBMsg test configuration for nRF54H20

ICBMSG test configuration for nRF54H20DK.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
tests: ipc_sessions: Update the nrf54h20_cpuapp_cpuppr configuration

Update nrf54h20_cpuapp_cpuppr configuration to use XIP PPR variant.
This variant can properly reboot itself making test valid with a cost of
increased power consumption and lower performance.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons commented:
I am not sure why I cannout use K_SPINLOCK together with atomic_cas operation on ARC architecture. It generates exception:
```
Running TESTSUITE bitpool
===================================================================
START - test_bitpool_atomic_cas
E: ***** Exception vector: 0x7, cause code: 0x0, parameter 0x0
E: Address 0x80008184
E: EV_PrivilegeV
E: Privilege violation
E:  r0: 0x80408ef4  r1: 0x80408f34  r2: 0x80408f74  r3: 0x1fe
E:  r4: 0x0  r5: 0x0  r6: 0x9f  r7: 0x0
E:  r8: 0x0  r9: 0x0 r10: 0x0 r11: 0x0
E: r12: 0xc r13: 0x80408f34  pc: 0x80008184
E:  blink: 0x80003938 status32: 0x8010209c
E: lp_end: 0x0 lp_start: 0x0 lp_count: 0x8040190c
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x804024e0 (test_bitpool_atomic_cas)
E: Halting system
```

Building command:
```
west build -p -b qemu_arc/qemu_arc_em
```
(from tests/kernel/common).

carlescufi commented:
> I am not sure why I cannout use K_SPINLOCK together with atomic_cas operation on ARC architecture. It generates exception:

Maybe  @evgeniy-paltsev, @abrodkin or @ruuddw know?

rakons commented:
Fixed bitpool module to work with atomic_val_t being 64bit value.

rakons commented:
Adding DNM label as I have to solve all the issues in bitpool first.
The PR with bitpool module only here: #86725

cfriedt commented:
@rakons - it might be worth it to keep the PR in draft until it's ready for review.

rakons committed:
ipc_service: ipc_icbmsg: Implementation of waititing threads counter

Limit the number of data release messages transferred by counting
number of threads waiting for data buffer.
Send the data release control message only if any thread is waiting
for the buffer on remote side.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons commented:
Compliance checks are false positive:

```
 BRACES: braces {} are required around if/while/for/else
File:include/zephyr/ipc/bitpool.h
Line:67
 BRACES: braces {} are required around if/while/for/else
File:include/zephyr/ipc/bitpool.h
Line:68
 BRACES: braces {} are required around if/while/for/else
File:include/zephyr/ipc/bitpool.h
Line:71
```
The construction here intentionally does not use braces {} around every for loop. Thanks to that we can use variables of different types inside for macro and still have a macro that executes its body that follows it in single braces {}.


```
 LINE_SPACING: Missing a blank line after declarations
File:tests/subsys/ipc/bitpool/src/bitpool.c
Line:45
 LINE_SPACING: Missing a blank line after declarations
File:tests/subsys/ipc/bitpool/src/bitpool.c
Line:97
 LINE_SPACING: Missing a blank line after declarations
File:tests/subsys/ipc/bitpool/src/bitpool.c
Line:167
 LINE_SPACING: Missing a blank line after declarations
File:tests/subsys/ipc/bitpool/src/bitpool.c
Line:207
 LINE_SPACING: Missing a blank line after declarations
File:tests/subsys/ipc/bitpool/src/bitpool.c
Line:247
```
Here macro ATOMIC_VAL_DEFINE is used to create a atomic bitpool variable. Checkpatch does not see it as a variable declaration thous returns an error. Technically I should get rid the error but it would be just for the sake of checkpatch pleasure and the code readability would drop.

rakons commented:
Now as I am writing the documentation I am wondering if using bitpool is really safe - It might generate an issues if the communicating cores are using different atomic_t sizes!

cfriedt reviewed (commented):
@rakons - iirc, there should be a way to add the macro to `.clang-format` to mitigate the false positives

cfriedt commented:
> Now as I am writing the documentation I am wondering if using bitpool is really safe - It might generate an issues if the communicating cores are using different atomic_t sizes!

That's a very distinct possibility.

An improvement in Zephyr would be to use standard C atomics which do support specific word sizes.

The main issue that this hasn't been done yet is that we support pre-c11 language standards. C11 atomics use generics (as do builtins), although we do use c11 standards elsewhere already.

https://en.cppreference.com/w/c/atomic

Compiler built-ins are also available as a backend (see `CONFIG_ATOMIC_OPERATIONS_BUILTIN`) and those also provide accessors for 8, 16, 32, and 64-bit types.

I think, eventually, it would make sense to use standard APIs, but the best / most stable path to getting there is TBD.

rakons committed:
ipc_service: ipc_icbmsg: Add optional status block alignment

Add optional alignment to the status block to reflect the fact
that cores to connect may use atomic_t variable of different sizes.
Optional force-atomic-size argument allow them to communicate
properly in such situation.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

rakons commented on include/zephyr/ipc/bitpool.h:
```diff
@@ -0,0 +1,468 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_SYS_bitpool_H_
+#define ZEPHYR_INCLUDE_SYS_bitpool_H_
```
Fixed manually, together with #endif at the end of the file.

rakons commented on include/zephyr/ipc/bitpool.h:
```diff
@@ -0,0 +1,468 @@
+#endif
+
+/**
+ * @brief Define an array of atomic values variables.
```
Fixed

rakons commented on include/zephyr/ipc/bitpool.h:
```diff
@@ -0,0 +1,468 @@
+ *
+ * @sa ATOMIC_DEFINE
+ *
+ * @param name Name of array of atomic variables.
```
Fixed

rakons commented:
@cfriedt - Thank you for the comments and review.
When it comes to atomic_t alignment issue I had added this commit: https://github.com/zephyrproject-rtos/zephyr/pull/86421/commits/79fe3fbca769511661a706cf7b3074bad2457fbd to handle the issue, being compatible with ANSI C99.
